Tests/test_SeqIO.py

```diff
@@ -435,15 +450,25 @@ def check_simple_write_read(records, indent=" "):
             # Beware of different quirks and limitations in the
             # valid character sets and the identifier lengths!
             if format in ["phylip", "phylip-sequential"]:
-                assert r1.id.replace("[", "").replace("]", "")[:10] == r2.id, \
-                    "'%s' vs '%s'" % (r1.id, r2.id)
+                assert PhylipIO.sanitize_name(r1.id, 10) == r2.id, \
+                        "'%s' vs '%s'" % (r1.id, r2.id)
             elif format == "phylip-relaxed":
-                assert r1.id.replace(" ", "").replace(':', '|') == r2.id, \
+                assert PhylipIO.sanitize_name(r1.id) == r2.id, \
                     "'%s' vs '%s'" % (r1.id, r2.id)
             elif format == "clustal":
                 assert r1.id.replace(" ", "_")[:30] == r2.id, \
                     "'%s' vs '%s'" % (r1.id, r2.id)
-            elif format in ["stockholm", "maf"]:
+            elif format == "stockholm":
+                r1_id = r1.id.replace(" ", "_")
+                if "start" in r1.annotations and "end" in r1.annotations:
```

Comment:
Can you expand on this, from the commit comment:

```
      Added a condition for the Stockholm parser for sequences that have
      a "start" and "end" annotation. The "stockholm" driver writes and
      parses these like "1ABC:A/X-Y", but the residue range is not
      included in the sequence ID when parsed from a structure.
```

Does something in your test cases trigger this? It feels like you may have found a Stockholm bug... but I remember this being complicated, so it could be deliberate.

Reply:
If a `SeqRecord` has a `start` and `end` annotation, `Bio.AlignIO.StockholmIO.StockholmWriter._write_record` [adds a `/start-end` suffix][1] to the sequence ID. When the Stockholm reader reads that record, it does not strip the suffix from the ID.

When a sequence is parsed using `cif-atom` or `pdb-atom`, the `start` and `end` annotations are set. So the round-trip test case looks like this:

1. `SeqRecord` parsed by `pdb-atom`: `id=1ABC:X`.
2. Sequence is written by the Stockholm writer: `id=1ABC:X/1-100`.
3. Sequence is *read* by the Stockholm reader: `id=1ABC:X/1-100`.
4. Ids of the `SeqRecords` are compared: `"1ABC:X" != "1ABC:X/1-100`.

My hack explicitly adds the range suffix to the original `SeqRecord`'s ID when checking the round trip with the Stockholm format. This looks like it *could* be a bug in the Stockholm reader, but it might be intentional. from `StockholmIO.StockholmIterator`:

> If an accession is provided for an entry in the meta data, IT WILL NOT
be used as the record.id (it will be recorded in the record's
annotations).  This is because some files have (sub) sequences from
different parts of the same accession (differentiated by different
start-end positions).

IMO, the current behaviour is unintuitive, but I don't know much about the Stockholm format. Maybe users expect the IDs to contain the range.

Minimal example:

``` python
>>> import io
  2 
  3 import Bio.SeqIO
  4 import Bio.SeqRecord
  5 import Bio.Seq
  6 
  7 record = Bio.SeqRecord.SeqRecord(Bio.Seq.Seq("A"), id="foo")
  8 record.annotations["start"] = 1
  9 record.annotations["end"] = 2
 10 buf = io.StringIO()
 11 Bio.SeqIO.write(record, buf, "stockholm")
 12 buf.seek(0)
 13 read = Bio.SeqIO.read(buf, "stockholm")
>>> record
SeqRecord(seq=Seq('A', Alphabet()), id='foo', name='<unknown name>', description='<unknown description>', dbxrefs=[])
>>> read
SeqRecord(seq=Seq('A', SingleLetterAlphabet()), id='foo/1-2', name='foo', description='<unknown description>', dbxrefs=[])
```

[1]: https://github.com/biopython/biopython/blob/153d4c67bf4a38caf65d5523ce1a152cdea010e6/Bio/AlignIO/StockholmIO.py#L251-L259